### PR TITLE
[lldb] Add frame recognizers for libc++ `std::invoke`

### PR DESCRIPTION
**Pull request by vogelsgesang**
With this commit, we also hide the implementation details of `std::invoke`. To do so, the `LibCXXFrameRecognizer` got a couple more regular expressions.

The regular expression passed into `AddRecognizer` became problematic, as it was evaluated on the demangled name. Those names also included result types for C++ symbols. For `std::__invoke` the return type is a huge `decltype(...)`, making the regular expresison really hard to write.

Instead, I added support to `AddRecognizer` for matching on the demangled names without result type and argument types.

By hiding the implementation details of `invoke`, also the back traces for `std::function` become even nicer, because `std::function` is using `__invoke` internally.

Co-authored-by: Adrian Prantl <aprantl@apple.com>

### CODE REVIEW

**Comment by vogelsgesang** [comment]
@mordante @ldionne FYI. Would be interested which other functions come to mind that should be hidden.

See https://github.com/llvm/llvm-project/pull/105457#issuecomment-2299996404 for screenshots of how this looks from a user's perspective

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp
Comment (jasonmolenda) [inline lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:58]: Love to see these example strings we're matching against

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp
Comment (vogelsgesang) [inline lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:71]: I wonder if we should simply hide everything starting with `^std::__[0-9]*::__.*`. Or are there any `__` functions in libc++ which are not implementation details and should be shown in stacktraces by default?

Comment (Michael137) [reply lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:71]: Agreed. Don't think we want to maintain this list.

(also, side-note, I think matching on `const` and `abi_tag` is probably redundant. Not a concern if we just filter out all the `std::__*::__` stuff).

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp
Comment (Michael137) [inline lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:108]:
```suggestion
        std::make_shared<RegularExpression>("^std::__[0-9]*::"),
```

File: lldb/source/Commands/Options.td
Comment (Michael137) [inline lldb/source/Commands/Options.td:1052]: Can just push this as an NFC separately

File: lldb/source/Target/StackFrameRecognizer.cpp
Comment (Michael137) [inline lldb/source/Target/StackFrameRecognizer.cpp:179]: Can't we just pass `symctx.GetFunctionName(entry.mangling_preference)` here right?

No need for the switch statement here and above and also no need to keep a map of `m_used_manglings`

Comment (vogelsgesang) [reply lldb/source/Target/StackFrameRecognizer.cpp:179]: I wasn't sure about performance here. The code previously computed the demanded name only once and then checked against all frame recognizers. I am not sure how expensive it is to demangle a name

Comment (Michael137) [reply lldb/source/Target/StackFrameRecognizer.cpp:179]: The demangling gets chached anyway. So wouldn't worry about it

Comment (vogelsgesang) [reply lldb/source/Target/StackFrameRecognizer.cpp:179]: By the way: I don't think I will be merging this part of my commit in the end. https://github.com/llvm/llvm-project/pull/105756 is very similar, and I think I will just rebase on it, after it got merged

Comment (adrian-prantl) [reply lldb/source/Target/StackFrameRecognizer.cpp:179]: @vogelsgesang I closed my PR, feel free to copy&paste what you find useful.

Comment (vogelsgesang) [reply lldb/source/Target/StackFrameRecognizer.cpp:159]: > The demangling gets chached anyway (in the Mangled class). So wouldn't worry about it

ah, great! I didn't know this. I removed the additional caching from `StackFrameRecognizer` again

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp
Comment (petrhosek) [reply lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:71]: libc++ ABI namespace doesn't have to be just a number, it can be any [string that starts with `__`](https://github.com/llvm/llvm-project/blob/c557d8520413476221a4f3bf2b7b3fed17681691/libcxx/CMakeLists.txt#L197). For example, in some of our Google projects we use `__ktl` and `__pw`. [libFuzzer uses `__Fuzzer`](https://github.com/llvm/llvm-project/blob/52ae891036e3ab1f668eb103c46ca57257901c6b/compiler-rt/lib/fuzzer/CMakeLists.txt#L166) in its internal libc++ build.

**Review by Michael137 — APPROVED** [review]
LGTM (with some minor comments). I'd wait for @adrian-prantl to also review this

File: lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp
Comment (vogelsgesang) [reply lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp:78]: right. I posted this comment before I saw your fix on `main`. By now, the commit contains `^std::__[^:]*::*

File: lldb/source/Plugins/SystemRuntime/MacOSX/AbortWithPayloadFrameRecognizer.cpp
Comment (vogelsgesang) [reply lldb/source/Plugins/SystemRuntime/MacOSX/AbortWithPayloadFrameRecognizer.cpp:201]: My editor auto-formats the whole file on save. This tends to be rather useful in other projects, but apparently not in LLVM

**Review by adrian-prantl — APPROVED** [review]
Thanks!

**Review by petrhosek — APPROVED** [review]

**Comment by petrhosek** [comment]
@vogelsgesang Can we please merge this PR to unbreak the builders that were broken by #104523?

File: lldb/test/API/commands/frame/recognizer/TestFrameRecognizer.py
Comment (vogelsgesang) [reply lldb/test/API/commands/frame/recognizer/TestFrameRecognizer.py:312]: Unfortunately, I was not able to write such a test case. Note that the `frame recognizer add` command does not yet expose a way to match on anything else than demangled names. I agree that we should be adding test cases here, as soon as `frame recognizer add` supports choosing the symbol mangling.

@adrian-prantl Are you planning to extend `frame recognizer add` accordingly?

**Comment by adrian-prantl** [comment]
Looks like this is failing on the Darwin bot: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/10463/

File: lldb/test/API/commands/frame/recognizer/TestFrameRecognizer.py
Comment (adrian-prantl) [reply lldb/test/API/commands/frame/recognizer/TestFrameRecognizer.py:312]: > Unfortunately, I was not able to write such a test case. Note that the `frame recognizer add` command does not yet expose a way to match on anything else than demangled names. I agree that we should be adding test cases here, as soon as `frame recognizer add` supports choosing the symbol mangling.
> 
> @adrian-prantl Are you planning to extend `frame recognizer add` accordingly?

It's not on my immediate todo list, but it would be an obvious next step.

**Comment by vogelsgesang** [comment]
fixed in #106281

**Pull request merged**
